### PR TITLE
[gen] Update the baseline forbidden configuration file on `Pos*W`.

### PR DESCRIPTION
**Pull request by ShaleXIONG**
We remove relaxation that is stronger than `Pos*W` in the baseline configuration file, which generates smaller test suites.
This removed edges includes:
- Fence on same location and end with write, e.g. `DMB.SY*W`.
- `Data`, `Addr`, and `Ctrl` dependency, with or without `Csel`, on the same location and end with write, e.g. `DpDatasW`.
- We remove some relaxations that `diy` can automatically generate.

### CODE REVIEW

File: gen/libdir/forbidden.conf [hunk 1]
```diff
@@ -113,13 +119,19 @@
...
+#-safe DpDataCselsW
 
 # let pob = pick-ctrl-dep; [Exp & W | HU | TLBI | DC.CVAU | IC]
--safe DpCtrlCseldW DpCtrlCselsW
+-safe DpCtrlCseldW
+# `DpCtrlCselsW` is stronger than `Pos*W` hence we emit the former.
+#-safe DpCtrlCselsW
 
 # let pob = pick-addr-dep; [Exp & M]; po; [Exp & W | HU]
 -safe [DpAddrCseldR PodRW] [DpAddrCseldR PosRW] [DpAddrCseldW PodWW] [DpAddrCseldW PosWW] [DpAddrCselsR PodRW] [DpAddrCselsR PosRW] [DpAddrCselsW PodWW] [DpAddrCselsW PosWW]
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: Do we need to change this as well?
```suggestion
-safe [DpAddrCseldR PodRW] [DpAddrCseldR PosRW] [DpAddrCseldW PodWW] [DpAddrCseldW PosWW] [DpAddrCselsR PodRW]  [DpAddrCselsW PodWW]
```

File: gen/libdir/forbidden.conf [hunk 2]
```diff
@@ -113,13 +119,19 @@
 #
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: Do we need to change:
`-safe [DpCtrldR ISB] [DpCtrldW ISB] [DpCtrlsR ISB] [DpCtrlsW ISB]`
to 
`-safe [DpCtrldR ISB] [DpCtrldW ISB] [DpCtrlsR ISB]` as well?

and 
`-safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB] [DpCtrlCselsW ISB]`
to
`-safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB]`

and
`-safe [DpAddrdR ISBd**] [DpAddrdR ISBs**] [DpAddrdW ISBd**] [DpAddrdW ISBs**] [DpAddrsR ISBd**] [DpAddrsR ISBs**] [DpAddrsW ISBd**] [DpAddrsW ISBs**]`
to
`-safe [DpAddrdR ISBd**] [DpAddrdR ISBs**] [DpAddrdW ISBd**] [DpAddrdW ISBs**] [DpAddrsR ISBd**] [DpAddrsR ISBs**] [DpAddrsW ISBd**]`

and
`-safe [DpAddrCseldR ISBd**] [DpAddrCseldR ISBs**] [DpAddrCseldW ISBd**] [DpAddrCseldW ISBs**] [DpAddrCselsR ISBd**] [DpAddrCselsR ISBs**] [DpAddrCselsW ISBd**] [DpAddrCselsW ISBs**]`
to
`-safe [DpAddrCseldR ISBd**] [DpAddrCseldR ISBs**] [DpAddrCseldW ISBd**] [DpAddrCseldW ISBs**] [DpAddrCselsR ISBd**] [DpAddrCselsR ISBs**] [DpAddrCselsW ISBd**]`

File: gen/libdir/forbidden.conf [hunk 3]
```diff
@@ -128,23 +140,35 @@
...
--safe DSB.LDdR* DSB.LDsR*
+-safe DSB.LDdR* DSB.LDsRR
+# The `DSB.LDsRW` is stronger than `Pos*W` hence we emit the former.
+#-safe DSB.LDsRW
 
 # let DSB-ob = [Exp & W]; po; [dsb.st]; po; [~(Imp & TTD & M | Imp & Instr & R)]
--safe DSB.STdW* DSB.STsW*
+-safe DSB.STdW* DSB.STsWR
+# The `DSB.STsWW` is stronger than `Pos*W` hence we emit the former.
+#-safe DSB.STsWW
```
Comment (relokin) [inline gen/libdir/forbidden.conf:193]: Do we need to change:
```
-safe [A PodR*] [A PosR*] [Q PodR*] [Q PosR*]
# An acquire atomic operation followed by a memory access also matches above
-safe [A Amo.Swp PodW*] [A Amo.Swp PosW*]
-safe [A Amo.Cas PodW*] [A Amo.Cas PosW*]
-safe [Q Amo.Swp PodW*] [Q Amo.Swp PosW*]
-safe [Q Amo.Cas PodW*] [Q Amo.Cas PosW*]
```
to
```
-safe [A PodR*] [A PosRR] [Q PodR*] [Q PosRR]
# An acquire atomic operation followed by a memory access also matches above
-safe [A Amo.Swp PodW*] [A Amo.Swp PosWR]
-safe [A Amo.Cas PodW*] [A Amo.Cas PosWR]
-safe [Q Amo.Swp PodW*] [Q Amo.Swp PosWR]
-safe [Q Amo.Cas PodW*] [Q Amo.Cas PosWR]
```

Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:193]: Yes, you are right. Let me update.

File: gen/libdir/forbidden.conf (hunk 2 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: I think regarding this list, we need to remove if the two relaxations are all in the same locations, if it is stronger than `[PosRW PosWW]`. Let me update this.

File: gen/libdir/forbidden.conf (hunk 3 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:193]: Also removed `[Pos*R Amo L]`.

File: gen/libdir/forbidden.conf (hunk 2 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: `[DpAddrCselsW ISBs**]` should be `[DpAddrCselsW ISBs*R]` where the read is still interested ? Since here is a read-write-read events.

File: gen/libdir/forbidden.conf (hunk 1 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: Should we keep the `[DpAddrCselsR PosRW]` but only remove `[DpAddrCselsW PosWW]`? The former is a read-read-write events, while the latter is a read-write-write events. This means the latter matches two `Pos*W`.

Comment (relokin) [reply gen/libdir/forbidden.conf]: I don't think it makes any difference in both cases the write is a local write successor of the first read as they are to the same location.

File: gen/libdir/forbidden.conf (hunk 2 above)
Comment (relokin) [reply gen/libdir/forbidden.conf]: Ah sorry I misread the last two cases (I was still thinking of ISB as a pseudo-edge), here is the corrected list:
Do we need to change:
`-safe [DpCtrldR ISB] [DpCtrldW ISB] [DpCtrlsR ISB] [DpCtrlsW ISB]`
to 
`-safe [DpCtrldR ISB] [DpCtrldW ISB] [DpCtrlsR ISB]` as well?

and 
`-safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB] [DpCtrlCselsW ISB]`
to
`-safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB]`

and
`-safe [DpAddrdR ISBd**] [DpAddrdR ISBs**] [DpAddrdW ISBd**] [DpAddrdW ISBs**] [DpAddrsR ISBd**] [DpAddrsR ISBs**] [DpAddrsW ISBd**] [DpAddrsW ISBs**]`
to (corrected) - the intention is to leave out cases that match the pattern `[DpAddr{s,d}* ISBs*W]` and cases that match the pattern `[DpAddrs** ISBs*W]`:
`-safe [DpAddrdR ISBdR*] [DpAddrdR ISBsRR] [DpAddrdW ISBdW*] [DpAddrdW ISBsWR] [DpAddrsR ISBdR*] [DpAddrsR ISBsRR] [DpAddrsW ISBdW*] [DpAddrsW ISBsWR]`


and 
`-safe [DpAddrCseldR ISBd**] [DpAddrCseldR ISBs**] [DpAddrCseldW ISBd**] [DpAddrCseldW ISBs**] [DpAddrCselsR ISBd**] [DpAddrCselsR ISBs**] [DpAddrCselsW ISBd**] [DpAddrCselsW ISBs**]`
to (corrected) - the intention is to leave out cases that match `[DpAddrCsel{s,d}W ISBsWW]` and cases that match `[DpAddrCsels** ISBs*W]`:
`-safe [DpAddrCseldR ISBdR*] [DpAddrCseldR ISBsR*] [DpAddrCseldW ISBdW*] [DpAddrCseldW ISBsW*] [DpAddrCselsR ISBdR*] [DpAddrCselsR ISBsRR] [DpAddrCselsW ISBdW*] [DpAddrCselsW ISBsWR]`

File: gen/libdir/forbidden.conf (hunk 1 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: I see. I will remove them all together.

File: gen/libdir/forbidden.conf (hunk 2 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: Cool. I will update accordingly.

File: gen/libdir/forbidden.conf [hunk 4]
```diff
@@ -75,19 +75,27 @@
...
 # let dob = ctrl; [Exp & W | HU | TLBI | DC.CVAU | IC]
 # mcat2config7 aarch64.cat -let dob
--safe DpCtrldW DpCtrlsW
+-safe DpCtrldW
+# `DpCtrlsW` is stronger than `Pos*W`.
+#-safe DpCtrlsW
 
 # let dob = addr; [Exp & M]; po; [Exp & W | HU]
--safe [DpAddrdW PodWW] [DpAddrdW PosWW] [DpAddrsW PodWW] [DpAddrsW PosWW] [DpAddrdR PodRW] [DpAddrdR PosRW] [DpAddrsR PodRW] [DpAddrsR PosRW]
+-safe [DpAddrdW PodWW] [DpAddrdW PosWW] [DpAddrsW PodWW] [DpAddrdR PodRW] [DpAddrdR PosRW] [DpAddrsR PodRW]
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: I think you can also remove `[DpAddrdW PosWW]` because DpAddrdW is already included in line 84 and PosWW is already included in line 73. Same for `[DpAddrdR PosRW]`.
```suggestion
-safe [DpAddrdW PodWW] [DpAddrsW PodWW] [DpAddrdR PodRW] [DpAddrsR PodRW]
```

File: gen/libdir/forbidden.conf [hunk 5]
```diff
@@ -100,51 +108,80 @@
...
 # let IFB-ob = [Exp & R]; pick-ctrl-dep; [IFB]; po
 # let IFB = ISB | EXC-ENTRY-IFB | EXC-RET-IFB
--safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB] [DpCtrlCselsW ISB]
+-safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB]
+# [DpCtrlCselsW ISB] is stronger than `Pos*W`
+#-safe [DpCtrlCselsW ISB]
 
 # let IFB-ob = [Exp & R]; addr; [Exp & M]; po; [IFB]; po
--safe [DpAddrdR ISBd**] [DpAddrdR ISBs**] [DpAddrdW ISBd**] [DpAddrdW ISBs**] [DpAddrsR ISBd**] [DpAddrsR ISBs**] [DpAddrsW ISBd**] [DpAddrsW ISBs**]
+-safe [DpAddrdR ISBdR*] [DpAddrdR ISBsR*] [DpAddrdW ISBdW*] [DpAddrdW ISBsW*] [DpAddrsR ISBdR*] [DpAddrsR ISBsRR] [DpAddrsW ISBdW*] [DpAddrsW ISBsWR]
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: Because `DpAddrd*` and `Pos*W` are already included, I think `[DpAddrd* ISBs*W]` is unnecessary:
```suggestion
-safe [DpAddrdR ISBdR*] [DpAddrdR ISBsRR] [DpAddrdW ISBdW*] [DpAddrdW ISBsWR] [DpAddrsR ISBdR*] [DpAddrsR ISBsRR] [DpAddrsW ISBdW*] [DpAddrsW ISBsWR]
```

File: gen/libdir/forbidden.conf [hunk 6]
```diff
@@ -100,51 +108,80 @@
...
+# `[DpAddrsW ISBsWW] [DpAddrsR ISBsRW]` are stronger than `Pos*W` as the former are a read followed by a write to the same location.
+#-safe [DpAddrsW ISBsWW] [DpAddrsR ISBsRW]
+
 # let IFB-ob = [Exp & R]; pick-addr-dep; [Exp & M]; po; [IFB]; po
--safe [DpAddrCseldR ISBd**] [DpAddrCseldR ISBs**] [DpAddrCseldW ISBd**] [DpAddrCseldW ISBs**] [DpAddrCselsR ISBd**] [DpAddrCselsR ISBs**] [DpAddrCselsW ISBd**] [DpAddrCselsW ISBs**]
-#
+-safe [DpAddrCseldR ISBdR*] [DpAddrCseldR ISBsR*] [DpAddrCseldW ISBdW*] [DpAddrCseldW ISBsW*] [DpAddrCselsR ISBdR*] [DpAddrCselsR ISBsRR] [DpAddrCselsW ISBdW*] [DpAddrCselsW ISBsWR]
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: Because DpAddrCseldW and Pos*W are already included, I think [DpAddrCseldW ISBsWW] is unnecessary
```suggestion
-safe [DpAddrCseldR ISBdR*] [DpAddrCseldR ISBsR*] [DpAddrCseldW ISBdW*] [DpAddrCseldW ISBsWR] [DpAddrCselsR ISBdR*] [DpAddrCselsR ISBsRR] [DpAddrCselsW ISBdW*] [DpAddrCselsW ISBsWR]
```

File: gen/libdir/forbidden.conf [hunk 7]
```diff
@@ -100,51 +108,80 @@
...
 
 # let pob = pick-ctrl-dep; [Exp & W | HU | TLBI | DC.CVAU | IC]
--safe DpCtrlCseldW DpCtrlCselsW
+-safe DpCtrlCseldW
+# `DpCtrlCselsW` is stronger than `Pos*W`.
+#-safe DpCtrlCselsW
 
 # let pob = pick-addr-dep; [Exp & M]; po; [Exp & W | HU]
--safe [DpAddrCseldR PodRW] [DpAddrCseldR PosRW] [DpAddrCseldW PodWW] [DpAddrCseldW PosWW] [DpAddrCselsR PodRW] [DpAddrCselsR PosRW] [DpAddrCselsW PodWW] [DpAddrCselsW PosWW]
+-safe [DpAddrCseldR PodRW] [DpAddrCseldR PosRW] [DpAddrCseldW PodWW] [DpAddrCseldW PosWW] [DpAddrCselsR PodRW] [DpAddrCselsW PodWW]
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: Because DpAddrCseldW and Pos*W are already included, I think [DpAddrCseldW PosWW] is unnecessary
```suggestion
-safe [DpAddrCseldR PodRW] [DpAddrCseldR PosRW] [DpAddrCseldW PodWW] [DpAddrCselsR PodRW] [DpAddrCselsW PodWW]
```

File: gen/libdir/forbidden.conf (hunk 4 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: Done.

File: gen/libdir/forbidden.conf (hunk 6 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: For this entire line, I change all the `ISB` to end with `R`. This is because `[DpAddrCsel** ISB**W]` is stronger than `[DpAddrCsel** Po**W]`. Some apply to `[DpAddr** ISB**W]`

File: gen/libdir/forbidden.conf (hunk 7 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: Done.

File: gen/libdir/forbidden.conf [hunk 8]
```diff
@@ -60,7 +60,7 @@
 #-relax {RELAX_LIST}
 
 
-# let ca = fr | co
```
Comment (relokin) [inline gen/libdir/forbidden.conf:63]: please revert this.

File: gen/libdir/forbidden.conf [hunk 9]
```diff
@@ -75,76 +75,119 @@
...
+# [DpCtrlCselsW ISB] is stronger than `Pos*W`
+#-safe [DpCtrlCselsW ISB]
 
 # let IFB-ob = [Exp & R]; addr; [Exp & M]; po; [IFB]; po
--safe [DpAddrdR ISBd**] [DpAddrdR ISBs**] [DpAddrdW ISBd**] [DpAddrdW ISBs**] [DpAddrsR ISBd**] [DpAddrsR ISBs**] [DpAddrsW ISBd**] [DpAddrsW ISBs**]
+-safe [DpAddrdR ISBdRR] [DpAddrdR ISBsRR] [DpAddrdW ISBdWR] [DpAddrsR ISBdRR] [DpAddrsR ISBsRR]
+# `[DpAddrsW ISBsWW] [DpAddrsR ISBsRW]` are stronger than `Pos*W` as the former are a read followed by a write to the same location.
+#-safe [DpAddrsW ISBsWW] [DpAddrsR ISBsRW]
+# `[DpAddr** ISB**W]` is stronger than `[DpAddr** Po**W]`; the latter can be combined by `diy`.
+#-safe [DpAddrdR ISBdRW] [DpAddrdR ISBsRW] [DpAddrdW ISBdWW] [DpAddrdW ISBsWW] [DpAddrsR ISBdRW] [DpAddrsW ISBdWW]
```
Comment (relokin) [inline gen/libdir/forbidden.conf:127]: Is this correct? I think we need 
`[DpAddrdR ISBdRW]`
`[DpAddrdW ISBdWW]`
`[DpAddrsR ISBdRW]`
`[DpAddrsW ISBdWW]`

And I am looking back at my comment:
> Because DpAddrd* and Pos*W are already included, I think [DpAddrd* ISBs*W] is unnecessary:

When you run tests and adding these as a `-safe` list to `diy7` and run them through herd7, did they print the Overlapping-lob flag?

File: gen/libdir/forbidden.conf [hunk 10]
```diff
@@ -75,76 +75,119 @@
...
+# `[DpAddrdW ISBsWR] [DpAddrsW ISBdWR]` are stronger than `[DpAddrdW PosWR] [DpAddrsW PodWR]` repsectively.
+#-safe [DpAddrdW ISBsWR] [DpAddrsW ISBdWR]
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: Similarly here, we need 
[DpAddrsW ISBdWR]

File: gen/libdir/forbidden.conf [hunk 11]
```diff
@@ -75,76 +75,119 @@
...
 # let IFB-ob = [Exp & R]; pick-ctrl-dep; [IFB]; po
 # let IFB = ISB | EXC-ENTRY-IFB | EXC-RET-IFB
--safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB] [DpCtrlCselsW ISB]
+-safe [DpCtrlCseldR ISB] [DpCtrlCseldW ISB] [DpCtrlCselsR ISB]
+# [DpCtrlCselsW ISB] is stronger than `Pos*W`
+#-safe [DpCtrlCselsW ISB]
 
 # let IFB-ob = [Exp & R]; addr; [Exp & M]; po; [IFB]; po
--safe [DpAddrdR ISBd**] [DpAddrdR ISBs**] [DpAddrdW ISBd**] [DpAddrdW ISBs**] [DpAddrsR ISBd**] [DpAddrsR ISBs**] [DpAddrsW ISBd**] [DpAddrsW ISBs**]
+-safe [DpAddrdR ISBdRR] [DpAddrdR ISBsRR] [DpAddrdW ISBdWR] [DpAddrsR ISBdRR] [DpAddrsR ISBsRR]
```
Comment (relokin) [inline gen/libdir/forbidden.conf]: I think this is better expressed as:
```suggestion
-safe [DpAddrsR ISBdR*] [DpAddrdR ISBdR*] [DpAddrsW ISBsWR] [DpAddrdW ISBdWR]
```
See my next comment as well, but the point is that `DpAddr**` is already included by line 84. So here we need to add all possible combinations of [DpAddr** ISB***] except the cases that have `ISBs*W` as this is already covered by Pos*W.

File: gen/libdir/forbidden.conf [hunk 12]
```diff
@@ -75,76 +75,119 @@
...
+#-safe DSB.SYs*W
 
 # let DSB-ob = [(Exp & R) \ NoRet | Imp & Tag & R]; po; [dsb.ld]; po; [~(Imp & TTD & M | Imp & Instr & R)]
--safe DSB.LDdR* DSB.LDsR*
+-safe DSB.LDdR* DSB.LDsRR
+# `DSB.LDsRW` is stronger than `Pos*W`.
+#-safe DSB.LDsRW
 
 # let DSB-ob = [Exp & W]; po; [dsb.st]; po; [~(Imp & TTD & M | Imp & Instr & R)]
 -safe DSB.STdW* DSB.STsW*
```
Comment (relokin) [inline gen/libdir/forbidden.conf]:
```suggestion
-safe DSB.STdW* DSB.STsWR
```

File: gen/libdir/forbidden.conf [hunk 13]
```diff
@@ -75,76 +75,119 @@
...
+#-safe [DpAddrdW ISBsWR] [DpAddrsW ISBdWR]
+
 # let IFB-ob = [Exp & R]; pick-addr-dep; [Exp & M]; po; [IFB]; po
--safe [DpAddrCseldR ISBd**] [DpAddrCseldR ISBs**] [DpAddrCseldW ISBd**] [DpAddrCseldW ISBs**] [DpAddrCselsR ISBd**] [DpAddrCselsR ISBs**] [DpAddrCselsW ISBd**] [DpAddrCselsW ISBs**]
-#
+-safe [DpAddrCseldR ISBdRR] [DpAddrCseldR ISBsRR] [DpAddrCseldW ISBdWR] [DpAddrCseldW ISBsWR] [DpAddrCselsR ISBdRR] [DpAddrCselsR ISBsRR] [DpAddrCselsW ISBdWR] [DpAddrCselsW ISBsWR]
+# `[DpAddrCselsW ISBsWW] [DpAddrCselsR ISBsRW]` are stronger than `Pos*W` as the former are a read followed by a write to the same location.
+#-safe [DpAddrCselsW ISBsWW] [DpAddrCselsR ISBsRW]
+# `[DpAddrCsel** ISB**W]` is stronger than `[DpAddrCsel** Po**W]`; the latter can be combined by `diy`.
+#-safe [DpAddrCseldR ISBdRW] [DpAddrCseldR ISBsRW] [DpAddrCseldW ISBdWW] [DpAddrCseldW ISBsWW] [DpAddrCselsR ISBdRW] [DpAddrCselsW ISBdWW]
```
Comment (relokin) [inline gen/libdir/forbidden.conf:136]: I think we need `[DpAddrCseldR ISBdRW] [DpAddrCseldR ISBsRW] [DpAddrCseldW ISBdWW] [DpAddrCselsR ISBdRW] [DpAddrCselsW ISBdWW]` in `-safe`

File: gen/libdir/forbidden.conf [hunk 14]
```diff
@@ -75,76 +75,119 @@
...
+#-safe [DpAddrsW ISBsWW] [DpAddrsR ISBsRW]
+# `[DpAddr** ISB**W]` is stronger than `[DpAddr** Po**W]`; the latter can be combined by `diy`.
+#-safe [DpAddrdR ISBdRW] [DpAddrdR ISBsRW] [DpAddrdW ISBdWW] [DpAddrdW ISBsWW] [DpAddrsR ISBdRW] [DpAddrsW ISBdWW]
+# `[DpAddrdW ISBsWR] [DpAddrsW ISBdWR]` are stronger than `[DpAddrdW PosWR] [DpAddrsW PodWR]` repsectively.
+#-safe [DpAddrdW ISBsWR] [DpAddrsW ISBdWR]
+
 # let IFB-ob = [Exp & R]; pick-addr-dep; [Exp & M]; po; [IFB]; po
--safe [DpAddrCseldR ISBd**] [DpAddrCseldR ISBs**] [DpAddrCseldW ISBd**] [DpAddrCseldW ISBs**] [DpAddrCselsR ISBd**] [DpAddrCselsR ISBs**] [DpAddrCselsW ISBd**] [DpAddrCselsW ISBs**]
-#
+-safe [DpAddrCseldR ISBdRR] [DpAddrCseldR ISBsRR] [DpAddrCseldW ISBdWR] [DpAddrCseldW ISBsWR] [DpAddrCselsR ISBdRR] [DpAddrCselsR ISBsRR] [DpAddrCselsW ISBdWR] [DpAddrCselsW ISBsWR]
```
Comment (relokin) [inline gen/libdir/forbidden.conf:132]: I think this is missing some relaxations. See my comment below as well.

I think you could write this as:
```suggestion
-safe [DpAddrCselsR ISBsRR] [DpAddrCselsR ISBdR*] [DpAddrCseldR ISBsR*] [DpAddrCseldR ISBdR*]
-safe [DpAddrCselsW ISBsWR] [DpAddrCselsW ISBdW*] [DpAddrCseldW ISBsWR] [DpAddrCseldW ISBdW*]
```

File: gen/libdir/forbidden.conf (hunk 8 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:63]: Sorry this is a rebase fault I will revert.

File: gen/libdir/forbidden.conf (hunk 9 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:127]: Relaxation `[DpAddrdR ISBdRW]`, which is `IFB-ob` is stronger than `[DpAddrdR PodRW]`, which is `dob` for example.

I run two tests here:
- a `diy7` test that the latter `[DpAddrdR PodRW]` is able to be combined by `diy7` itself. 
- I apply similar test here but run `flag empty (IFB-ob & dob) as Not-overlapping-lob`. The `~empty` sanctify check shows all flag, and the actually `empty` check show no flag.

File: gen/libdir/forbidden.conf (hunk 10 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: Sorry, you are right here. Also to this comment line, I run test on `[DpAddrdW ISBsWR]` against `flag empty (IFB-ob & dob) as Not-overlapping-lob` to confirmed it can be removed.

File: gen/libdir/forbidden.conf (hunk 11 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: I think it is the same question as before.

I really mean to eliminate the `DpAddr** ISB**W` in this line, because it overlapped with the `DpAddr** Po**W`. This is tested against `flag empty (IFB-ob & dob) as Not-overlapping-lob`.

File: gen/libdir/forbidden.conf (hunk 13 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:136]: Similar to the non-csel version.
I really mean to eliminate the `DpAddrCsel** ISB**W`, which is in `IFB-ob` in this line, because it overlapped with the `DpAddrCSel** Po**W`, which is in `pob`. This is tested against `flag empty (IFB-ob & pob) as Not-overlapping-lob`.

File: gen/libdir/forbidden.conf (hunk 14 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:132]: It follows the same answer:
Similar to the non-csel version. I really mean to eliminate the `DpAddrCsel** ISB**W`, which is in `IFB-ob` in this line, because it overlapped with the `DpAddrCSel** Po**W`, which is in `pob`. This is tested against `flag empty (IFB-ob & pob) as Not-overlapping-lob`.

File: gen/libdir/forbidden.conf (hunk 12 above)
Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf]: Done. Thanks for catch it. I shall be more careful on editing this file.

File: gen/libdir/forbidden.conf (hunk 9 above)
Comment (relokin) [reply gen/libdir/forbidden.conf:127]: I see, I thought we were only looking to avoid at overlaps with Pos*W.

File: gen/libdir/forbidden.conf (hunk 14 above)
Comment (relokin) [reply gen/libdir/forbidden.conf:132]: Ok thanks!

File: gen/libdir/forbidden.conf [hunk 15]
```diff
@@ -75,76 +75,119 @@
...
+-safe DpCtrldW
+# `DpCtrlsW` is stronger than `Pos*W`.
+#-safe DpCtrlsW
 
 # let dob = addr; [Exp & M]; po; [Exp & W | HU]
--safe [DpAddrdW PodWW] [DpAddrdW PosWW] [DpAddrsW PodWW] [DpAddrsW PosWW] [DpAddrdR PodRW] [DpAddrdR PosRW] [DpAddrsR PodRW] [DpAddrsR PosRW]
+-safe [DpAddrdW PodWW] [DpAddrsW PodWW] [DpAddrdR PodRW] [DpAddrsR PodRW]
+# `[DpAddrsW PosWW] [DpAddrsR PosRW]` are stronger than `Pos*W` as the former are a read followed by a write to the same location.
+# `[DpAddrdW PosWW] [DpAddrdR PosRW]` can be combined by `diy`
+#-safe [DpAddrsW PosWW] [DpAddrsR PosRW]
```
Comment (relokin) [inline gen/libdir/forbidden.conf:97]: This comment is missing:
```suggestion
#-safe [DpAddrsW PosWW] [DpAddrsR PosRW]  [DpAddrdW PosWW] [DpAddrdR PosRW]
```
I don't think you need to change this, although I think you might want to. It's more to make sure you tested the fact that tests with these relaxations would have overlapping lob edges.
For this you would need something like the following in the cat file:
```
assert empty ((addr; [Exp & M]; po; [Exp & W | HU]) & ((addr|lwfs)+)) as Overlapping-lob
```

Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:97]: Will update the comment. Lets me explore the assert check too.

Comment (ShaleXIONG) [reply gen/libdir/forbidden.conf:97]: [update]
Just to confirm the check gave us the expected behaviour.

A similar test check of `[DpAddrCseldW PosWW]` against `flag empty ((pick-addr-dep; [Exp & M]; po; [Exp & W | HU]) & ((pick-addr-dep|lwfs)+)) as Overlapping-pick-addr-lwfs` also gave the expected behaviour.

I think it is safe to remove those relaxations.
